lib/runtime/errors/transaction_validity.go

```diff
@@ -0,0 +1,105 @@
+// on how it relates to other transactions. It is a result of the form:
+// Result<ValidTransaction, TransactionValidityError>
+type TransactionValidityError scale.VaryingDataType
+
+var (
+	errInvalidType     = errors.New("invalid validity type")
+	errInvalidResult   = errors.New("invalid error value")
+	errInvalidTypeCast = errors.New("invalid type cast")
+)
+
+// Set will set a VaryingDataTypeValue using the underlying VaryingDataType
+func (tve *TransactionValidityError) Set(val scale.VaryingDataTypeValue) (err error) {
+	vdt := scale.VaryingDataType(*tve)
+	err = vdt.Set(val)
+	if err != nil {
+		return err
+	}
+	*tve = TransactionValidityError(vdt)
+	return nil
+}
+
+// Value will return the value from the underlying VaryingDataType
+func (tve *TransactionValidityError) Value() (val scale.VaryingDataTypeValue) {
+	vdt := scale.VaryingDataType(*tve)
+	return vdt.Value()
+}
+
+// Error will return the error underlying TransactionValidityError
+func (tve *TransactionValidityError) Error() string {
+	invalidTxn, ok := tve.Value().(InvalidTransaction)
+	if !ok {
+		unknownTxn, ok2 := tve.Value().(UnknownTransaction)
+		if !ok2 {
+			return errInvalidTypeCast.Error()
+		}
+		return unknownTxn.Error()
+	}
+	return invalidTxn.Error()
+}
+
+// NewTransactionValidityError is constructor for TransactionValidityError
+func NewTransactionValidityError() TransactionValidityError {
+	vdt, err := scale.NewVaryingDataType(NewInvalidTransaction(), NewUnknownTransaction())
+	if err != nil {
+		panic(err)
+	}
+	return TransactionValidityError(vdt)
+}
+
+var (
+	ErrInvalidTxn = errors.New("invalid transaction")
+	ErrUnknownTxn = errors.New("unknown transaction")
+)
+
+// UnmarshalTransactionValidity takes the result of the validateTransaction runtime call and unmarshalls it
+// TODO use custom result issue #2780
+func UnmarshalTransactionValidity(res []byte) (*transaction.Validity, error) {
+	validTxn := transaction.Validity{}
+	txnValidityErrResult := NewTransactionValidityError()
+	txnValidityResult := scale.NewResult(validTxn, txnValidityErrResult)
+	err := scale.Unmarshal(res, &txnValidityResult)
+	if err != nil {
+		return nil, fmt.Errorf("scale decoding transaction validity result: %w", err)
+	}
+	txnValidityRes, err := txnValidityResult.Unwrap()
+	if err != nil {
+		scaleWrappedErr, ok := err.(scale.WrappedErr)
+		if ok {
+			txnValidityErr, ok := scaleWrappedErr.Err.(TransactionValidityError)
+			if !ok {
+				return nil, fmt.Errorf("%w: %T", errInvalidTypeCast, scaleWrappedErr.Err)
+			}
+
+			switch txnValidityErr.Value().(type) {
+			// TODO use custom result issue #2780
+			case InvalidTransaction:
+				return nil, fmt.Errorf("%w: %s", ErrInvalidTxn, txnValidityErr.Error())
+			case UnknownTransaction: // do nothing
+				return nil, fmt.Errorf("%w: %s", ErrUnknownTxn, txnValidityErr.Error())
+			default:
+				panic(fmt.Sprintf("unsupported transaction validity error: %T", txnValidityErr.Value()))
+			}
```

Comment:
You should be using the value itself as the error since you already fulfilled the error interface on all the `VaryingDataTypeValue` types that ` TransactionValidityError` would have as a value.  

There are very few cases where a panic is appropriate.  I don't think this is one of those cases.

```suggestion
			switch val := txnValidityErr.Value().(type) {
			// TODO use custom result issue #2780
			case InvalidTransaction:
				return nil, val
			case UnknownTransaction: // do nothing
				return nil, val
			default:
				return nil, fmt.Errorf("unsupported transaction validity error: %T", txnValidityErr.Value())
			}
```

Reply:
- I suggested to use a sentinel error and stringify the error since it simplifies the calling code (just errors.Is instead of errors.As + vdt.Value() + type switch, so no need to know vdt/vdt types, just check go 1.13 sentinel errors)
- The `default:` case is a programming error. We should list all value types there even if we do nothing with them to ensure we don't forget one. That's one the panic is useful.

Reply:
We have first class types that represent the types in the substrate codebase.  Yes we will need to use `errors.As` to distinguish between error types, but what is wrong with that?  These types may contain extra data from the type that the calling code should be able to inspect on.

Reply:
> Yes we will need to use errors.As to distinguish between error types, but what is wrong with that? These types may contain extra data from the type that the calling code should be able to inspect on.

The VDT (from the top of my head) only contains a string anyway, so I would rather dumb down the Go API and keep the VDT as implementation detail in this case, and have the calling code use `errors.As`. I would rather change it to export the VDT when/if there is a need for it. We could also do this without breaking the Go API technically. But up to you guys, if you feel there is a need to have the vdt accessible by callers already, go for it!

Not gonna lie, it's also that Jimmy had this code exposing the vdt, and I told him we could do it with just errors and no exposed vdt, so I'm also feeling bad for suggesting something we want to revert now 😢 

> Why would the default case be programmer error? We only care about the two types encapsulated by the VDT, but there's no reason why parity couldn't add to this enum in the future. And why would we panic in this unexpected case?

Can IO (network) stuff in another type than the 2 we have in the VDT? If so, then yes a panic is not appropriate at all!
Otherwise, it means we are not handling all the VDT types in our code (we should add the `case` for each, even if we do nothing), so that's a programming error (not a user/IO error), and so we should panic. Most likely no one would pick up the error log and it will be much harder to debug in the future, than having a panic which, despite being fatal, is quite explicit.

Reply:
As a caller of this function, I would never expect this function to panic.  The only way I can see this case ever being hit, is if someone injected the value into the `txnValidityErr` VDT, which is very unlikely.  But even then, in this very unlikely case, we should leave up to the caller of the function to panic via exposing a sentinel error.

Reply:
If the only way this code path triggers is through a code change that doesn't exist yet and shouldn't exist, I think we should still panic. This is not due to unmarshalling/decoding bad data AFAIK. 🤔